# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1640810311 -3600
# Parent  549d9cc035bc9ee378a0a78d2376628d74056c2a
Feature: Multi-line tree follow-up: Fix scrollbar issues.

diff --git a/layout/xul/tree/nsTreeBodyFrame.cpp b/layout/xul/tree/nsTreeBodyFrame.cpp
--- a/layout/xul/tree/nsTreeBodyFrame.cpp
+++ b/layout/xul/tree/nsTreeBodyFrame.cpp
@@ -516,19 +516,25 @@ Maybe<CSSIntRegion> nsTreeBodyFrame::Get
 
   return Some(region);
 }
 
 nsresult nsTreeBodyFrame::Invalidate() {
   if (mUpdateBatchNest) return NS_OK;
 
   // Re-evaluate whether we're multi-line.
+  bool wasMultiline = mIsMultiline;
   nsCOMPtr<nsIMultilineTreeView> multilineView = do_QueryInterface(mView);
   if (multilineView) multilineView->IsMultiline(&mIsMultiline);
   mRowHeight = GetRowHeight();
+  // The correct page length is needed for the scrollbar update.
+  if (!mHasFixedRowCount)
+    mPageLength =
+        (mRowHeight > 0) ? (mInnerBox.height / mRowHeight) : mRowCount;
+  if (mIsMultiline != wasMultiline) FullScrollbarsUpdate(false);
 
   InvalidateFrame();
 
   return NS_OK;
 }
 
 nsresult nsTreeBodyFrame::InvalidateColumn(nsTreeColumn* aCol) {
   if (mUpdateBatchNest) return NS_OK;
